lib/auth/sessions.go

```diff
@@ -0,0 +1,125 @@
+/*
+Copyright 2020 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package auth
+
+import (
+	"context"
+	"time"
+
+	"github.com/gravitational/teleport"
+	"github.com/gravitational/teleport/lib/jwt"
+	"github.com/gravitational/teleport/lib/services"
+	"github.com/gravitational/teleport/lib/utils"
+	"github.com/gravitational/teleport/lib/wrappers"
+	"github.com/gravitational/trace"
+	"github.com/pborman/uuid"
+)
+
+// CreateAppWebSession creates and inserts a services.WebSession into the
+// backend with the identity of the caller used to generate the certificate.
+// The certificate is used for all access requests, which is where access
+// control is enforced.
+func (s *Server) CreateAppSession(ctx context.Context, req services.CreateAppSessionRequest, user services.User, checker services.AccessChecker) (services.WebSession, error) {
+	// Check that a matching parent web session exists in the backend.
+	parentSession, err := s.GetWebSession(req.Username, req.ParentSession)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// Don't let the TTL of the child certificate go longer than the parent.
+	ttl := checker.AdjustSessionTTL(parentSession.GetExpiryTime().Sub(s.clock.Now()))
+
+	// Create certificate for this session.
+	privateKey, publicKey, err := s.GetNewKeyPairFromPool()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	certs, err := s.generateUserCert(certRequest{
+		user:      user,
+		publicKey: publicKey,
+		checker:   checker,
+		ttl:       ttl,
+		// Set the login to be a random string. Application certificates are never
+		// used to log into servers but SSH certificate generation code requires a
+		// principal be in the certificate.
+		traits: wrappers.Traits(map[string][]string{
+			teleport.TraitLogins: []string{uuid.New()},
```

Comment:
I can split them out if you feel strongly about it, but right now all of our certificate generation logic assumes it will return both a SSH and x509 certificate.